### Reduction operation: the sum of the numbers in the range [0, a)

In [1]:
import time

def reduc_operation(a):
    """Compute the sum of the numbers in the range [0, a)."""
    x = 0
    for i in range(a):
        x += i
    return x

# Secuencial

value = 1000000

initialTime = time.time()
suma = reduc_operation(value)
finalTime = time.time()

print("Time taken by reduction operation:", (finalTime - initialTime), "seconds")

# Utilizando las operaciones mágicas de ipython
%timeit -r 2 reduc_operation(value)

print(f"\n \t Computing the sum of numbers in the range [0, value): {suma}\n")

Time taken by reduction operation: 0.3469116687774658 seconds
282 ms ± 12 ms per loop (mean ± std. dev. of 2 runs, 1 loop each)

 	 Computing the sum of numbers in the range [0, value): 499999500000



In [2]:
# 3.2 a)
import time

# Crear una lista de Python con 10^6 elementos (1,000,000)
my_list = list(range(1000000))

# Opción 1: Sumar con un bucle for
def for_loop_sum():
    total = 0
    for num in my_list:
        total += num
    return total

# Medición del tiempo para el bucle for
initialTime = time.time()
total_for = for_loop_sum()
finalTime = time.time()
print("Time taken by reduction operation:", (finalTime - initialTime), "seconds")

# Opción 2: Sumar usando la función sum()
def sum_function():
    return sum(my_list)

# Medición del tiempo para sum()
initialTime = time.time()
total_sum = sum_function()
finalTime = time.time()
print("Time taken by sum() operation:", (finalTime - initialTime), "seconds")

# Medición con %timeit
print("\nTiming with %timeit:")

# Medición con %timeit para el bucle for
%timeit -r 2 -n 1000 for_loop_sum()

# Medición con %timeit para la función sum()
%timeit -r 2 sum_function()

# Mostrar el resultado de la suma
print(f"\n\tComputing the sum of numbers in the range [0, 1000000]: {total_sum}\n")
import time

# Crear una lista de Python con 10^6 elementos
my_list = list(range(1000000))

# Opción 1: Sumar con un bucle for
def for_loop_sum():
    total = 0
    for num in my_list:
        total += num
    return total

# Opción 2: Sumar usando sum() de Python
def sum_function():
    return sum(my_list)

# Medición con time.time() para el bucle for
initialTime = time.time()
total_for = for_loop_sum()
finalTime = time.time()
print("Time taken by reduction operation (for loop):", (finalTime - initialTime), "seconds")

# Medición con time.time() para sum()
initialTime = time.time()
total_sum = sum_function()
finalTime = time.time()
print("Time taken by sum() operation:", (finalTime - initialTime), "seconds")

# Utilizando las operaciones mágicas de ipython
print("\nTiming with %timeit:")

# Medición con %timeit para el bucle for
%timeit -r 2 -n 1000 for_loop_sum()

# Medición con %timeit para la función sum()
%timeit -r 2 sum_function()

# Mostrar el resultado de las sumas de todas las maneras posibles
print(f"\n\tComputing the sum of numbers in the range [0, 1000000]:")
print(f"\tSum using for loop: {total_for}")
print(f"\tSum using sum(): {total_sum}")

Time taken by reduction operation: 0.2133791446685791 seconds
Time taken by sum() operation: 0.033614397048950195 seconds

Timing with %timeit:
206 ms ± 475 µs per loop (mean ± std. dev. of 2 runs, 1,000 loops each)
30 ms ± 56.1 µs per loop (mean ± std. dev. of 2 runs, 10 loops each)

	Computing the sum of numbers in the range [0, 1000000]: 499999500000

Time taken by reduction operation (for loop): 0.20290756225585938 seconds
Time taken by sum() operation: 0.03325963020324707 seconds

Timing with %timeit:
206 ms ± 34.8 µs per loop (mean ± std. dev. of 2 runs, 1,000 loops each)
30.1 ms ± 178 µs per loop (mean ± std. dev. of 2 runs, 10 loops each)

	Computing the sum of numbers in the range [0, 1000000]:
	Sum using for loop: 499999500000
	Sum using sum(): 499999500000


In [3]:
#3.2 b)
import time
import numpy as np

# Convertir la lista a un array de NumPy
my_array = np.array(my_list)

# Opción 1: Sumar con un bucle for
def for_loop_sum_array():
    total = 0
    for num in my_array:
        total += num
    return total

# Medición con time.time() para el bucle for
initialTime = time.time()
total_array_for = for_loop_sum_array()
finalTime = time.time()
print("Time taken by reduction operation with array:", (finalTime - initialTime), "seconds")

# Opción 2: Sumar usando np.sum() de NumPy
def sum_function_array():
    return np.sum(my_array)

# Medición con time.time() para np.sum()
initialTime = time.time()
total_array_sum = sum_function_array()
finalTime = time.time()
print("Time taken by np.sum() operation with array:", (finalTime - initialTime), "seconds")

# Medición con %timeit
print("\nTiming with %timeit:")

# Medición con %timeit para el bucle for
%timeit -r 2 -n 1000 for_loop_sum_array()

# Medición con %timeit para la función np.sum()
%timeit -r 2 sum_function_array()


Time taken by reduction operation with array: 0.6505894660949707 seconds
Time taken by np.sum() operation with array: 0.001982450485229492 seconds

Timing with %timeit:
646 ms ± 6.26 ms per loop (mean ± std. dev. of 2 runs, 1,000 loops each)
1.35 ms ± 7.41 µs per loop (mean ± std. dev. of 2 runs, 1,000 loops each)


Los resultados muestran que el bucle for es el enfoque más lento para realizar la suma, lo que refleja la sobrecarga de Python al manejar operaciones repetitivas de manera explícita. Aunque este método es directo, su rendimiento es limitado cuando se trabaja con grandes cantidades de datos, ya que la iteración manual no está optimizada. Por otro lado, la función sum() de Python mejora considerablemente el tiempo de ejecución gracias a su implementación optimizada en C, permitiendo una suma mucho más rápida de los elementos de la lista. Finalmente, np.sum() de NumPy ofrece un rendimiento similar al de sum(), pero su verdadera ventaja radica en su capacidad para manejar arrays más grandes y multidimensionales de forma más eficiente, aprovechando las optimizaciones propias de NumPy. Esto demuestra que usar funciones nativas optimizadas o bibliotecas especializadas como NumPy puede acelerar significativamente el procesamiento en comparación con enfoques más simples como los bucles for.

In [11]:
from numba import njit
import numpy as np
import time

# Versión original con NumPy (para comparar)
def numpy_sum_original(arr):
    return np.sum(arr)

# Versiones optimizadas con Numba
@njit
def numpy_sum_numba_optimized(arr):
    """Suma con np.sum() + Numba"""
    return np.sum(arr)

@njit
def manual_sum_numba(arr):
    """Suma con bucle for + Numba"""
    x = 0
    for i in arr:
        x += i
    return x

# Configuración del array (optimizado para Numba)
value = 10**6
array = np.arange(value, dtype=np.int64)  # dtype explícito para mejor rendimiento

# ---- Comparativa de tiempos ----
print("=== Comparativa NumPy vs. Numba (1 millón de elementos) ===")

# 1. NumPy puro
start = time.time()
result_numpy = numpy_sum_original(array)
numpy_time = time.time() - start
print(f"\nNumPy puro: {numpy_time:.6f} segundos")

# 2. Numba + np.sum() (incluyendo compilación JIT)
start = time.time()
result_numba_sum = numpy_sum_numba_optimized(array)  # Primera ejecución (compila)
numba_sum_time = time.time() - start
print(f"\nNumba + np.sum() - Primera ejecución: {numba_sum_time:.6f} segundos")

start = time.time()
result_numba_sum = numpy_sum_numba_optimized(array)  # Segunda ejecución (ya compilado)
numba_sum_time_compiled = time.time() - start
print(f"Numba + np.sum() - Segunda ejecución: {numba_sum_time_compiled:.6f} segundos")

# 3. Numba + bucle manual
start = time.time()
result_manual_numba = manual_sum_numba(array)  # Primera ejecución (compila)
numba_manual_time = time.time() - start
print(f"\nNumba + bucle for - Primera ejecución: {numba_manual_time:.6f} segundos")

start = time.time()
result_manual_numba = manual_sum_numba(array)  # Segunda ejecución (ya compilado)
numba_manual_time_compiled = time.time() - start
print(f"Numba + bucle for - Segunda ejecución: {numba_manual_time_compiled:.6f} segundos")

# ---- Resultados ----
print("\n=== Resumen de resultados ===")
print(f"Suma NumPy puro: {result_numpy}")
print(f"Suma Numba + np.sum(): {result_numba_sum}")
print(f"Suma Numba + bucle for: {result_manual_numba}")

# Speedup comparativo
speedup_sum = numpy_time / numba_sum_time_compiled
speedup_manual = numpy_time / numba_manual_time_compiled
print(f"\nSpeedup Numba+np.sum() vs NumPy: {speedup_sum:.1f}x")
print(f"Speedup Numba+bucle vs NumPy: {speedup_manual:.1f}x")

=== Comparativa NumPy vs. Numba (1 millón de elementos) ===

NumPy puro: 0.001886 segundos

Numba + np.sum() - Primera ejecución: 0.444284 segundos
Numba + np.sum() - Segunda ejecución: 0.001507 segundos

Numba + bucle for - Primera ejecución: 0.448630 segundos
Numba + bucle for - Segunda ejecución: 0.002103 segundos

=== Resumen de resultados ===
Suma NumPy puro: 499999500000
Suma Numba + np.sum(): 499999500000
Suma Numba + bucle for: 499999500000

Speedup Numba+np.sum() vs NumPy: 1.3x
Speedup Numba+bucle vs NumPy: 0.9x


Explicación de los resultados obtenidos por Numba (3.3 b)

Los resultados muestran diferencias significativas en los tiempos de ejecución entre los métodos evaluados. NumPy demostró ser el más eficiente para la operación de suma, completando el cálculo en 0.0019 segundos, gracias a su implementación optimizada en C.

La versión con Numba presentó un comportamiento particular: durante la primera ejecución requirió 0.44 segundos debido al proceso de compilación JIT, pero en ejecuciones posteriores redujo su tiempo a 0.0015 segundos, superando levemente el rendimiento de NumPy. Este patrón confirma que Numba ofrece mejores resultados en usos repetitivos tras la compilación inicial.

En cuanto al enfoque con bucle manual optimizado con Numba, alcanzó un tiempo de 0.0021 segundos, demostrando que puede aproximarse al rendimiento de las operaciones vectorizadas de NumPy, especialmente en casos donde no es posible utilizar funciones pre-optimizadas.

Todos los métodos produjeron el resultado correcto (499,999,500,000), validando su precisión numérica. Estas observaciones sugieren que la elección entre NumPy y Numba debe basarse en el contexto específico de aplicación, considerando tanto la naturaleza del cálculo como el patrón de uso.
